### ISSUE TITLE
Explore using react-navigation native-stack on web as well 

### ISSUE THREAD

**Issue by mountiny** [Expensify/App#38694]
### Problem

To achieve 100% cross-platform code and user exeprience, we thrive to use the same code and technology on all platforms. Not all technology is ready for such cross-platform use or it simply does not want to be designed that way. However, we should look into way to limit a code platform-fragmentation as much as possible as its improving the code quality and readability too.

In this case, we are looking into using the native-stack in native platforms to gain from the native feel and potential performance boost. That leaves us with the react-navigation/stack for web.

### Solution

Let's explore what is missing in the native-stack, that App might need in its web implementation. 

Then once we have better idea of what effort it would need to make native-stack the only stack we use in the App, we can discuss with the maintainers, if its something they are interested in and would be supportive of. If yes, let's do it and unify the code base.

**Comment by kirillzyusko** [comment]
I can work on this ticket 🙌

**Comment by kirillzyusko** [comment]
# Introduction 👋 

In this post I'm trying to describe:
- what we need to have in `native-stack` to be able to use it on web (to have cross platform code for navigation);
- potential ways for achieving this

## ⁉️ 🔑 Key aspects preventing `native-stack` from being used on web

Here is a preliminary report of what needs to be implemented in `native-stack` before we can start to use it on web 👀 

### ⚠️ Missing props from specification

#### `animation`

Right now you can specify any animations and transitions will not be animated, because code in `native-stack` is not taking this property into consideration.

I think it was made with a purpose, because otherwise you'll have to replicate a lot of functionality/animations/transitions on web, but the idea behind `native-stack` navigator was to re-use native navigation primitives.

Anyway, this is a critical point, since without animated transitions application on web loses all attractiveness 😔 

#### `animationTypeForReplace`

I think it is not implemented right now because there is no animations at all on web platform.

But we are using this property, so it should be implemented along with `animation` property.

### ⛔ Missing navigation events

All events on `web` are not getting dispatched (I think mainly because of the fact, that all transitions are not animated, so you can use plain `useEffect` to detect a moment when screen is fully visible).

The code in Expensify is heavily relies on transition events, such as `transitionEnd` etc. and without these events app is not working at all (infinite loading `Search` and other screens)

### 📚 Properties that we rely on

#### `cardStyle`

We are actively using this property and do some modification of this property depends on screen size. Right now without this property our app is totally broken on web, because a list of chat can not be displayed because parent size is null, and we never mount our `Sidebar` component, and because of that messages can not be loaded.

So this is a very critical property for Expensify app to work properly 😅 

### 🎣 Hooks 

For `Overlay` component we are using `useCardAnimation` to get animation progress and then do a custom interpolation (changing `opacity` property).

## 📝 Proposed design

Below I have two proposals in my mind that we can follow for implementation.

### 1️⃣ Expose navigation primitives from `react-native-screens`

Taking all design concepts into consideration I think the best way to implement this feature would be to expose "native" primitives from `react-native-screens` on web.

I understand, that there is no navigation primitives on web, but I think what we can do is to create an API that will have an identical API of what we have at the moment in `react-native-screens` and try to replicate such behavior on web (i. e. implement all `animation` variations, expose `onAppear` and other callbacks).

### 2️⃣ Replicate all logic in `native-stack`

The second approach (which I think is less desirable because it'll break design principles) is to implement all these transitions on `native-stack` level (thus we'll not add a new code to core `react-native-screens` package and instead of that polyfilled functionality will be fully in `native-stack` package).

## PoC implementation

I did all my experiments in https://github.com/Expensify/App/commit/f59566739f42bf2aa868b0f4bcd8bff64a533562

It's not fully working application, but it contains changes that allows to run app at least.

<hr>

@mountiny whom we should ask to review it? 👀

**Comment by mountiny** [comment]
Thank you @kirillzyusko I think that is a great summary. We can ask the react-navigation maintainers if they would consider these changes

**Comment by kirillzyusko** [comment]
@mountiny yep, that would be the next step 👀

**Comment by mountiny** [comment]
I have asked SWM for now

**Comment by roryabraham** [comment]
ok, thanks for looking into this but given that web support for native-stack is not a priority for react-navigation, I think we should close this now.

**Comment by WoLewicki** [comment]
I can see that it has been closed but I think that both of the proposed solutions would end up being a copy of the `stack` implementation since we would not want to invent the wheel. Or am I missing some context @kirillzyusko ?

**Comment by roryabraham** [comment]
If it were my project, I'd happily extract common logic between `stack` and `native-stack` to be reused on web where appropriate, to provide better feature parity across platforms in `native-stack`. But if we just have to shim to use `stack` on web and `native-stack` on mobile, that's fine for now
